### PR TITLE
[scheduler] Include depth and function descriptor in scheduling class

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR adds task depth to the scheduling class.

Every task's depth is 1 greater than its owner's depth. The depth of an actor task is the same as that of its actor creation task.

Giving different depths their own scheduling class allows us to easily queue and schedule tasks of different depths separately (i.e. prevent deadlock in a recursive function call with something like #17887) 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/task/task_spec.h [hunk 1]
```diff
@@ -29,11 +29,52 @@
...
+    std::stringstream buffer;
+    buffer << "{"
+           << "depth=" << depth << " "
+           << "function_descriptor=" << function_descriptor->ToString() << " "
+           << "resource_set="
+           << "{";
+    for (const auto &pair : resource_set.GetResourceMap()) {
+      buffer << pair.first << " : " << pair.second << ", ";
+    }
+    buffer << "}}";
```
Comment (ericl) [inline src/ray/common/task/task_spec.h:58]: Do we show these in debug_state.txt (number of active tasks per scheduling class)?

**Review by ericl — COMMENTED** [review]
How can we add a unit test for this?

**Comment by wuisawesome** [comment]
Actually it looks like this can't be split out. It looks like it fundamentally won't pass `test_basic_3.py::test_fair_queuing` without the rest of #17887.

**Comment by ericl** [comment]
> Actually it looks like this can't be split out. It looks like it fundamentally won't pass test_basic_3.py::test_fair_queuing without the rest of #17887.

Can we do it feature flagged? (can set them to no-ops with the flag off). I'd like to see the comments addressed above prior to moving forward.

File: src/ray/common/task/task_spec.h (hunk 1 above)
Comment (wuisawesome) [reply src/ray/common/task/task_spec.h:58]: In this PR, we don't track the number of active tasks per scheduling class, but this debug string will be used once we do track it.

**Review by ericl — CHANGES_REQUESTED** [review]
Can we add an integration test that tests the depth is incremented properly for recursive tasks?

**Comment by iycheng** [comment]
> Actually it looks like this can't be split out. It looks like it fundamentally won't pass `test_basic_3.py::test_fair_queuing` without the rest of #17887.

Any reason for this one? I think this PR only adds more info right? I mean nothing changed actually.

**Comment by wuisawesome** [comment]
> Any reason for this one? I think this PR only adds more info right? I mean nothing changed actually.

I think `test_fair_queuing` currently works because the default iteration order happens to go through the queues in an order that works. Changing the scheduling class changes the order that we iterate/dispatch off the queues. (I believe before the different depth tasks were interlaced on a single queue, whereas with this PR we would always go through the `f` queue first).

File: python/ray/tests/test_scheduling.py [hunk 2]
```diff
@@ -674,6 +675,48 @@ def train(self):
...
+    start_infeasible.remote()
+    infeasible.remote()
+
+    # We expect the 2 calls to `infeasible` to be separate scheduling classes
+    # because one has depth=1, and the other has depth=2.
+
+    metric_name = "ray_internal_num_infeasible_scheduling_classes"
+
+    def condition():
+        _, metric_names, metric_samples = fetch_prometheus([prom_addr])
```
Comment (ericl) [inline python/ray/tests/test_scheduling.py:710]: nice test! This is a great pattern we can use in the future for other challenging cases.

File: src/ray/common/ray_config_def.h [hunk 3]
```diff
@@ -103,6 +103,11 @@ RAY_CONFIG(bool, preallocate_plasma_memory, false)
 /// even balancing of load. Low values (min 0.0) encourage more load spreading.
 RAY_CONFIG(float, scheduler_spread_threshold, 0.5);
 
+/// Whether to include function descriptors, and depth in the scheduling class.
+/// This causes tasks to be queued differently, so it may effect scheduling
+/// behavior.
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:108]: Probably can add a TODO to remove this once fully enabled.

File: python/ray/tests/test_scheduling.py [hunk 4]
```diff
@@ -674,6 +675,48 @@ def train(self):
...
+        if metric_name in metric_names:
+            for sample in metric_samples:
+                if sample.name == metric_name and sample.value == 2.0:
```
Comment (ericl) [inline python/ray/tests/test_scheduling.py:713]: Btw could we also test going to 3 and 4?

**Review by ericl — APPROVED** [review]
LGTM with minor comments.

**Comment by wuisawesome** [comment]
Tests look good. Windows failure is `test_cancel` which is flakey/failing on master.

**Pull request merged**